com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerVisualizer.cs

```diff
@@ -0,0 +1,183 @@
+        }
+
+        protected void OnDisable()
+        {
+            if (controllerDetectedAction == null) { return; }
+            controllerDetectedAction.Disable();
+            controllerDetectedAction.started -= RenderControllerVisuals;
+            controllerDetectedAction.canceled -= RemoveControllerVisuals;
+        }
+        private void RenderControllerVisuals(InputAction.CallbackContext context)
+        {
+            RenderControllerVisuals(context.control.device);
+        }
+
+        private void RenderControllerVisuals(UnityInputSystem.InputDevice inputDevice)
+        {
+            // This process may change in the future as unity updates its input subsystem.
+            // In the future, there will be a different way of distinguishing between phsyical controllers
+            // and tracked hands, forgoing the UnityEngine.XR.InputDevices route
+
+            // Upon detecting a generic input device with the appropriate usages, load or remove the controller visuals
+            // when appropriate
+            if (inputDevice is UnityInputSystem.XR.XRController xrInputDevice && xrInputDevice.usages.Contains(targetUsage))
+            {
+                // Fallback visuals are only used if NO hand joints are detected
+                // OR the input device is specifically a simulated controller that is in the MotionController Simulation Mode.
+                bool useFallbackVisuals;
+                bool isSimulatedController;
+                if (xrInputDevice is MRTKSimulatedController simulatedController)
+                {
+                    useFallbackVisuals = simulatedController.SimulationMode == ControllerSimulationMode.MotionController;
+                    isSimulatedController = true;
+                }
+                else
+                {
+                    useFallbackVisuals = !HandsUtils.GetSubsystem().TryGetJoint(TrackedHandJoint.Palm, handNode, out _);
+                    isSimulatedController = false;
+                }
+                InstantiateControllerVisuals(!isSimulatedController, useFallbackVisuals);
+            }
+        }
+
+        // Private reference to the gameobject which represents the visualized controller
+        // Needs to be explicitly set to null in cases where no controller visuals are ever loaded.
+        private GameObject ControllerGameObject = null;
+
+        /// <summary>
+        /// Tries to instantiate controller visuals for the specified hand node.
+        /// </summary>
+        /// <param name="usePlatformVisuals">Whether or not to try to load visuals from the platform provider</param>
+        /// <param name="useFallbackVisuals">Whether or not to use the fallback controller visuals</param>
+        private async void InstantiateControllerVisuals(bool usePlatformVisuals, bool useFallbackVisuals)
+        {
+            // Disable any pre-existing controller models before trying to render new ones.
+            platformLoadedGameObject?.SetActive(false);
```

Comment:
I don't think null propagators are safe to use on GameObjects, right? Or is it only MonoBehaviours?

Reply:
Any `UnityEngine.Object` :( so yeah, not on GOs either

Reply:
More info: https://blog.unity.com/technology/custom-operator-should-we-keep-it

Reply:
Fixed